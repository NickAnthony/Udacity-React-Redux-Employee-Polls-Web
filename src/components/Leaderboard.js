import { connect } from "react-redux";
import UserAvatarPicture from "./UserAvatarPicture";

const Leaderboard = ({ sortedUsers, users }) => {
  return (
    <div className="container">
      <table>
        <tr className="leaderboard-header">
          <th>User</th>
          <th>Answered</th>
          <th>Created</th>
        </tr>
        {sortedUsers.map((sortedUser) => {
          const username = sortedUser[0];
          const user = users[username];
          return (
            <tr key={username}>
              <td className="profile-container">
                <UserAvatarPicture avatarURL={user.avatarURL} size={40} />
                <div className="profile-name-container">
                  <span className="profile-name">{user.name}</span>
                  <span className="profile-username">{username}</span>
                </div>
              </td>
              <td>{user.numAnswers}</td>
              <td>{user.numQuestions}</td>
            </tr>
          );
        })}
      </table>
    </div>
  );
};

const mapStateToProps = ({ users }) => {
  // First, we have to calculate each
  const unsortedUsers = Object.keys(users).map((username) => {
    const user = {
      ...users[username],
    };
    user["numAnswers"] = Object.keys(user.answers).length;
    user["numQuestions"] = user.questions.length;
    return [username, user.numAnswers, user.numQuestions];
  });
  // This function sorts usernames by the number of questions they've answered
  // then by the number of questions they asked.
  const sortedUsers = unsortedUsers.sort((a, b) => {
    if (a[1] === b[1]) {
      return b[2] - a[2];
    } else {
      return b[1] - a[1];
    }
  });
  return {
    sortedUsers: sortedUsers,
    users,
  };
};

export default connect(mapStateToProps)(Leaderboard);
