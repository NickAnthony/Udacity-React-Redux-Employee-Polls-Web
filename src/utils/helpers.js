import { useLocation, useNavigate, useParams } from "react-router-dom";

// Copied directly from https://reactrouter.com/en/main/start/faq
export function withRouter(Component) {
  function ComponentWithRouterProp(props) {
    let location = useLocation();
    let navigate = useNavigate();
    let params = useParams();
    return <Component {...props} router={{ location, navigate, params }} />;
  }

  return ComponentWithRouterProp;
}

export function formatTimestamp(timestamp) {
  const date = new Intl.DateTimeFormat("en-US", {
    year: "numeric",
    month: "numeric",
    day: "numeric",
  }).format(timestamp);
  const time = new Intl.DateTimeFormat("en-US", {
    hour: "numeric",
    minute: "2-digit",
  }).format(timestamp);
  return `${time} | ${date}`;
}
