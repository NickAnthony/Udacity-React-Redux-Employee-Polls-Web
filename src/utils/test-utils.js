import { configureStore } from "@reduxjs/toolkit";
import reducers from "../reducers";
import middleware from "../middleware";
import { users, questions } from "../utils/_DATA";
import CreateQuestion from "../components/CreateQuestion";
import CreateUser from "../components/CreateUser";
import Dashboard from "../components/Dashboard";
import Leaderboard from "../components/Leaderboard";
import Login from "../components/Login";
import Question from "../components/Question";
import UserProfile from "../components/UserProfile";
import { Provider } from "react-redux";
import { MemoryRouter, Route, Routes } from "react-router-dom";

export function createTestStore(initialState) {
  const store = configureStore({
    reducer: reducers,
    middleware: middleware,
    preloadedState: initialState,
  });
  return store;
}

export function createTestStoreWithInitialData() {
  return createTestStore({
    users,
    questions,
    authedUser: "nickanthony",
  });
}

export function createTestAppWithStore(initialRoute) {
  return (
    <Provider store={createTestStoreWithInitialData()}>
      <MemoryRouter initialEntries={[initialRoute]}>
        <Routes>
          <Route path="/poll/:id" element={<Question />} />
        </Routes>
      </MemoryRouter>
    </Provider>
  );
}
