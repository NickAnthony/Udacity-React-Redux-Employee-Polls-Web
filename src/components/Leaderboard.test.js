import { render } from "@testing-library/react";
import { createTestAppWithStore } from "../utils/test-utils";
import Leaderboard from "./Leaderboard";

describe("Leaderboard", () => {
  it("will match the snapshot for the Leaderboard.", () => {
    var component = render(createTestAppWithStore(<Leaderboard />));
    expect(component).toMatchSnapshot();
  });

  it("will populate the leaderboard table with numbers.", () => {
    var component = render(createTestAppWithStore(<Leaderboard />));
    // Verify we have 5 rows in the leaderboard
    expect(component.getAllByTestId("user-row").length).toEqual(5);
    const numberAnsweredComponent = component.getAllByTestId(
      "user-number-answered"
    )[0];
    expect(numberAnsweredComponent.textContent).toEqual("4");
    const numberCreatedComponent = component.getAllByTestId(
      "user-number-created"
    )[0];
    expect(numberAnsweredComponent.textContent).toEqual("4");
  });
});
