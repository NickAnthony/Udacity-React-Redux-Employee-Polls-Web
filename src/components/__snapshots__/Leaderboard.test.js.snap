// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard will match the snapshot for the Leaderboard. 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <table>
          <tbody>
            <tr
              class="leaderboard-header"
            >
              <th>
                User
              </th>
              <th>
                Answered
              </th>
              <th>
                Created
              </th>
            </tr>
            <tr
              data-testid="user-row"
            >
              <td
                class="profile-container"
              >
                <div
                  class="profile-avatar"
                  style="background-image: url(DefaultAvatar.png); height: 40px; width: 40px;"
                />
                <div
                  class="profile-name-container"
                >
                  <span
                    class="profile-name"
                  >
                    Sarah Edo
                  </span>
                  <span
                    class="profile-username"
                  >
                    sarahedo
                  </span>
                </div>
              </td>
              <td
                data-testid="user-number-answered"
              >
                4
              </td>
              <td
                data-testid="user-number-created"
              >
                2
              </td>
            </tr>
            <tr
              data-testid="user-row"
            >
              <td
                class="profile-container"
              >
                <div
                  class="profile-avatar"
                  style="background-image: url(DefaultAvatar.png); height: 40px; width: 40px;"
                />
                <div
                  class="profile-name-container"
                >
                  <span
                    class="profile-name"
                  >
                    Mike Tsamis
                  </span>
                  <span
                    class="profile-username"
                  >
                    mtsamis
                  </span>
                </div>
              </td>
              <td
                data-testid="user-number-answered"
              >
                3
              </td>
              <td
                data-testid="user-number-created"
              >
                2
              </td>
            </tr>
            <tr
              data-testid="user-row"
            >
              <td
                class="profile-container"
              >
                <div
                  class="profile-avatar"
                  style="background-image: url(DefaultAvatar.png); height: 40px; width: 40px;"
                />
                <div
                  class="profile-name-container"
                >
                  <span
                    class="profile-name"
                  >
                    Tyler McGinnis
                  </span>
                  <span
                    class="profile-username"
                  >
                    tylermcginnis
                  </span>
                </div>
              </td>
              <td
                data-testid="user-number-answered"
              >
                2
              </td>
              <td
                data-testid="user-number-created"
              >
                2
              </td>
            </tr>
            <tr
              data-testid="user-row"
            >
              <td
                class="profile-container"
              >
                <div
                  class="profile-avatar"
                  style="background-image: url(DefaultAvatar.png); height: 40px; width: 40px;"
                />
                <div
                  class="profile-name-container"
                >
                  <span
                    class="profile-name"
                  >
                    Zenobia Oshikanlu
                  </span>
                  <span
                    class="profile-username"
                  >
                    zoshikanlu
                  </span>
                </div>
              </td>
              <td
                data-testid="user-number-answered"
              >
                1
              </td>
              <td
                data-testid="user-number-created"
              >
                0
              </td>
            </tr>
            <tr
              data-testid="user-row"
            >
              <td
                class="profile-container"
              >
                <div
                  class="profile-avatar"
                  style="background-image: url(https://i.redd.it/ofmlty7h86fa1.jpg); height: 40px; width: 40px;"
                />
                <div
                  class="profile-name-container"
                >
                  <span
                    class="profile-name"
                  >
                    Nick Anthony
                  </span>
                  <span
                    class="profile-username"
                  >
                    nickanthony
                  </span>
                </div>
              </td>
              <td
                data-testid="user-number-answered"
              >
                1
              </td>
              <td
                data-testid="user-number-created"
              >
                0
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <table>
        <tbody>
          <tr
            class="leaderboard-header"
          >
            <th>
              User
            </th>
            <th>
              Answered
            </th>
            <th>
              Created
            </th>
          </tr>
          <tr
            data-testid="user-row"
          >
            <td
              class="profile-container"
            >
              <div
                class="profile-avatar"
                style="background-image: url(DefaultAvatar.png); height: 40px; width: 40px;"
              />
              <div
                class="profile-name-container"
              >
                <span
                  class="profile-name"
                >
                  Sarah Edo
                </span>
                <span
                  class="profile-username"
                >
                  sarahedo
                </span>
              </div>
            </td>
            <td
              data-testid="user-number-answered"
            >
              4
            </td>
            <td
              data-testid="user-number-created"
            >
              2
            </td>
          </tr>
          <tr
            data-testid="user-row"
          >
            <td
              class="profile-container"
            >
              <div
                class="profile-avatar"
                style="background-image: url(DefaultAvatar.png); height: 40px; width: 40px;"
              />
              <div
                class="profile-name-container"
              >
                <span
                  class="profile-name"
                >
                  Mike Tsamis
                </span>
                <span
                  class="profile-username"
                >
                  mtsamis
                </span>
              </div>
            </td>
            <td
              data-testid="user-number-answered"
            >
              3
            </td>
            <td
              data-testid="user-number-created"
            >
              2
            </td>
          </tr>
          <tr
            data-testid="user-row"
          >
            <td
              class="profile-container"
            >
              <div
                class="profile-avatar"
                style="background-image: url(DefaultAvatar.png); height: 40px; width: 40px;"
              />
              <div
                class="profile-name-container"
              >
                <span
                  class="profile-name"
                >
                  Tyler McGinnis
                </span>
                <span
                  class="profile-username"
                >
                  tylermcginnis
                </span>
              </div>
            </td>
            <td
              data-testid="user-number-answered"
            >
              2
            </td>
            <td
              data-testid="user-number-created"
            >
              2
            </td>
          </tr>
          <tr
            data-testid="user-row"
          >
            <td
              class="profile-container"
            >
              <div
                class="profile-avatar"
                style="background-image: url(DefaultAvatar.png); height: 40px; width: 40px;"
              />
              <div
                class="profile-name-container"
              >
                <span
                  class="profile-name"
                >
                  Zenobia Oshikanlu
                </span>
                <span
                  class="profile-username"
                >
                  zoshikanlu
                </span>
              </div>
            </td>
            <td
              data-testid="user-number-answered"
            >
              1
            </td>
            <td
              data-testid="user-number-created"
            >
              0
            </td>
          </tr>
          <tr
            data-testid="user-row"
          >
            <td
              class="profile-container"
            >
              <div
                class="profile-avatar"
                style="background-image: url(https://i.redd.it/ofmlty7h86fa1.jpg); height: 40px; width: 40px;"
              />
              <div
                class="profile-name-container"
              >
                <span
                  class="profile-name"
                >
                  Nick Anthony
                </span>
                <span
                  class="profile-username"
                >
                  nickanthony
                </span>
              </div>
            </td>
            <td
              data-testid="user-number-answered"
            >
              1
            </td>
            <td
              data-testid="user-number-created"
            >
              0
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
